File: calculator.js
import { OPERATORS, FUNCTIONS, CONSTANTS } from './operators.js';

const Calculator = (function() {
    // 1. 预处理模块 - 处理属性调用和运算符生成
    function preprocess(expr, operators, functions) {
        // 动态添加属性调用运算符
        for (const [name, func] of Object.entries(FUNCTIONS)) {
            if (func.asProperty) {
                operators.add('.' + name);
                OPERATORS['.' + name] = {
                    precedence: 3,
                    args: 1,
                    func: func.func,
                    position: 'postfix'
                };
            }
        }

        // 按长度降序排列运算符，确保先匹配较长的运算符
        const sortedOperators = new Set([...operators].sort((a, b) => b.length - a.length));

        return { expr, operators: sortedOperators };
    }

    // 预处理辅助函数
    function collectFunctionName(expr, start) {
        let funcName = '';
        let i = start;
        while (i < expr.length && /[a-zA-Z]/.test(expr[i])) {
            funcName += expr[i];
            i++;
        }
        return { funcName, length: i - start };
    }

    function collectPrecedingExpr(expr, start) {
        let depth = 0;
        let i = start;

        // 跳过空白字符
        while (i >= 0 && /\s/.test(expr[i])) i--;

        if (i >= 0) {
            if (expr[i] === ')') {
                depth = 1;
                i--;
                while (i >= 0 && depth > 0) {
                    if (expr[i] === '(') depth--;
                    if (expr[i] === ')') depth++;
                    i--;
                }
                while (i >= 0 && /[a-zA-Z0-9]/.test(expr[i])) i--;
            } else {
                while (i >= 0 && /[a-zA-Z0-9]/.test(expr[i])) i--;
            }
        }

        return {
            start: i,
            expr1: expr.slice(0, i + 1),
            expr2: expr.slice(i + 1, start + 1)
        };
    }

    // 2. 词法分析模块
    function tokenize(expr, operators, functions, constants) {
        const tokens = [];
        let i = 0;
        expr = expr.replace(/\s/g, '');

        // 从 OPERATORS 中获取所有分隔符
        const delimiters = new Set(
            Object.entries(OPERATORS)
                .filter(([_, op]) => op.type === 'delimiter')
                .map(([key]) => key)
        );

        // 从 OPERATORS 中获取所有分隔符
        const separators = new Set(
            Object.entries(OPERATORS)
                .filter(([_, op]) => op.type === 'separator')
                .map(([key]) => key)
        );

        function isUnaryMinus() {
            // 负号的条件：
            // 1. 是表达式开始
            // 2. 前一个token是左括号
            // 3. 前一个token是运算符
            // 4. 前一个token是分隔符(逗号)
            return tokens.length === 0 || 
                   (tokens[tokens.length-1][0] === 'operator' && tokens[tokens.length-1][1] === '(') ||
                   (tokens[tokens.length-1][0] === 'operator' && tokens[tokens.length-1][1] !== ')') ||
                   tokens[tokens.length-1][0] === 'separator';
        }

        function collectString() {
            let str = '';
            
            while (i < expr.length) {
                const char = expr[i];
                const remainingExpr = expr.slice(i);
                
                // 检查是否是运算符
                let isOperator = false;
                for (const op of operators) {
                    if (remainingExpr.startsWith(op)) {
                        isOperator = true;
                        break;
                    }
                }
                if (isOperator) break;

                // 检查是否是分隔符或定界符
                if (delimiters.has(char) || separators.has(char)) break;

                str += char;
                i++;
            }
            return str;
        }

        function collectIdentifier() {
            let name = '';
            while (i < expr.length && /[a-zA-Z0-9_]/.test(expr[i])) {
                name += expr[i];
                i++;
            }
            return name;
        }

        while (i < expr.length) {
            const char = expr[i];

            // 检查运算符
            let foundOperator = false;
            for (const op of operators) {
                if (expr.startsWith(op, i)) {
                    // 特殊处理减号，区分一元负号和二元减号
                    if (op === '-') {
                        if (isUnaryMinus()) {
                            tokens.push(['operator', 'unary-']);  // 使用一元负号
                        } else {
                            tokens.push(['operator', '-']);       // 使用二元减号
                        }
                    } else {
                        tokens.push(['operator', op]);
                    }
                    i += op.length;
                    foundOperator = true;
                    break;
                }
            }
            if (foundOperator) continue;

            // 检查函数名
            if (/[a-zA-Z]/.test(char)) {
                const name = collectIdentifier();
                if (functions.has(name)) {
                    tokens.push(['function', name]);
                } else if (constants.has(name)) {
                    tokens.push(['constant', name]);
                } else {
                    // 不是函数或常量，作为普通字符串处理
                    tokens.push(['string', name]);
                }
                continue;
            }

            // 处理分隔符和定界符
            if (separators.has(char)) {
                tokens.push(['separator', char]);
            } else if (delimiters.has(char)) {
                tokens.push(['delimiter', char]);
            } else {
                // 其他所有情况都作为字符串处理
                const str = collectString();
                if (str) tokens.push(['string', str]);
                continue;
            }

            i++;
        }

        return tokens;
    }

    // 3. 语法分析模块
    function buildAst(tokens, operators, functions) {
        let current = 0;
        const maxDepth = 100;

        function createNode(value, args = []) {
            return { value, args };
        }

        function checkRecursionDepth(depth) {
            if (depth > maxDepth) {
                throw new Error('表达式过于复杂或可能存在无限递归');
            }
        }

        function parseExpression(depth = 0) {
            checkRecursionDepth(depth);
            return parseByPrecedence(1, depth + 1);  // 从最低优先级开始
        }

        // 按优先级解析
        function parseByPrecedence(precedence, depth = 0) {
            checkRecursionDepth(depth);
            
            // 如果是最高优先级，解析一元运算符
            if (precedence === 4) {
                return parseUnary(depth + 1);
            }

            let left = parseByPrecedence(precedence + 1, depth + 1);

            while (current < tokens.length) {
                const [type, token] = tokens[current];
                
                // 检查是否是当前优先级的中缀运算符
                if (type !== 'operator' || 
                    !OPERATORS[token] || 
                    OPERATORS[token].position !== 'infix' ||
                    OPERATORS[token].precedence !== precedence) {
                    break;
                }

                current++;
                const right = parseByPrecedence(
                    // 对于右结合运算符（如 **），使用同级递归
                    token === '**' ? precedence : precedence + 1, 
                    depth + 1
                );
                left = createNode(token, [left, right]);
            }

            return left;
        }

        function parseUnary(depth = 0) {
            checkRecursionDepth(depth);
            const [type, token] = tokens[current];
            
            // 处理前缀运算符
            if (type === 'operator' && 
                OPERATORS[token]?.position === 'prefix') {
                current++;
                const operand = parseUnary(depth + 1);
                return createNode(token, [operand]);
            }

            return parsePostfix(depth + 1);
        }

        function parsePostfix(depth = 0) {
            checkRecursionDepth(depth);
            let left = parsePrimary(depth + 1);

            while (current < tokens.length) {
                const [type, token] = tokens[current];
                
                // 处理后缀运算符
                if (type !== 'operator' || 
                    !OPERATORS[token] || 
                    OPERATORS[token].position !== 'postfix') {
                    break;
                }
                
                current++;
                left = createNode(token, [left]);
            }

            return left;
        }

        function parsePrimary(depth = 0) {
            checkRecursionDepth(depth);
            const [type, token] = tokens[current];
            current++;

            // 处理字符串（包括数字）
            if (type === 'string') {
                return createNode(token);
            }

            // 处理常量
            if (type === 'constant') {
                return createNode(token);
            }

            if (type === 'function') {
                const args = parseFunctionArgs(depth + 1);
                return createNode(token, args);
            }

            if (type === 'operator' && token === '(') {
                const expr = parseExpression(depth + 1);
                if (current >= tokens.length || 
                    tokens[current][0] !== 'operator' || 
                    tokens[current][1] !== ')') {
                    throw new Error('括号不匹配');
                }
                current++;  // 跳过右括号
                return expr;
            }

            throw new Error(`意外的token: ${token}`);
        }

        function parseFunctionArgs(depth = 0) {
            checkRecursionDepth(depth);
            const args = [];
            
            // 检查左括号
            if (tokens[current][0] !== 'operator' || tokens[current][1] !== '(') {
                throw new Error('函数调用缺少左括号');
            }
            current++;  // 跳过左括号

            // 空参数列表
            if (tokens[current][0] === 'operator' && tokens[current][1] === ')') {
                current++;
                return args;
            }

            // 解析参数列表
            while (true) {
                args.push(parseExpression(depth + 1));

                if (tokens[current][0] === 'operator' && tokens[current][1] === ')') {
                    current++;
                    break;
                }

                if (tokens[current][0] !== 'separator' || tokens[current][1] !== ',') {
                    throw new Error('函数参数格式错误');
                }
                current++;  // 跳过逗号
            }

            return args;
        }

        // 开始解析
        const ast = parseExpression(0);
        if (current < tokens.length) {
            throw new Error('表达式解析未完成');
        }
        return ast;
    }

    // 4. 求值模块
    function evaluate(node, operators, functions) {
        if (!node) return 0;

        // 处理字符串节点，尝试转换为数值
        if (node.type === 'string') {
            const str = node.value;
            
            // 处理大数
            if (str.includes('e') || str.includes('E')) {
                return parseFloat(str);
            }

            // 尝试转换为数值
            const num = parseFloat(str);
            if (!isNaN(num)) return num;

            // 如果无法转换为数值，保持字符串
            return str;
        }

        // 处理常量
        if (node.type === 'constant') {
            return CONSTANTS[node.value];
        }

        const args = node.args.map(arg => evaluate(arg, operators, functions));

        if (operators.has(node.value)) {
            return OPERATORS[node.value].func(...args);
        }

        if (functions.has(node.value)) {
            const func = FUNCTIONS[node.value].func;
            return node.value === 'max' || node.value === 'min' ?
                args.length === 1 ? func(args) : func(...args) :
                func(...args);
        }

        throw new Error(`未知的操作符或函数: ${node.value}`);
    }

    // 5. 辅助函数
    function isStartOfNumber(tokens) {
        return !tokens.length || 
            (tokens[tokens.length-1][0] === 'operator' && 
             tokens[tokens.length-1][1] === '(') ||
            (tokens[tokens.length-1][0] === 'separator');
    }

    function generateMermaidAST(node, counter = {count: 0}) {
        if (!node) return { nodes: '', edges: '' };

        const currentId = `node${counter.count++}`;
        let nodes = `${currentId}["${typeof node.value === 'number' ? 
            node.value.toString() : 
            node.value}"];\n`;
        let edges = '';

        for (let i = 0; i < node.args.length; i++) {
            const child = node.args[i];
            const childResult = generateMermaidAST(child, counter);
            nodes += childResult.nodes;
            edges += childResult.edges;
            edges += `${currentId} -->|"arg${i + 1}"| node${counter.count-1};\n`;
        }

        return { nodes, edges };
    }

    // 6. 公共API
    return {
        calculate(expr) {
            const operators = new Set(Object.keys(OPERATORS));
            const functions = new Set(Object.keys(FUNCTIONS));
            const constants = new Set(Object.keys(CONSTANTS));

            // 预处理时动态添加运算符
            const { expr: processedExpr, operators: sortedOperators } = preprocess(expr, operators, functions);
            const tokens = tokenize(processedExpr, sortedOperators, functions, constants);
            const ast = buildAst(tokens, operators, functions);
            return evaluate(ast, operators, functions);
        },

        getASTNode(expr) {
            const operators = new Set(Object.keys(OPERATORS));
            const functions = new Set(Object.keys(FUNCTIONS));
            const constants = new Set(Object.keys(CONSTANTS));

            const { expr: processedExpr, operators: sortedOperators } = preprocess(expr, operators, functions);
            const tokens = tokenize(processedExpr, sortedOperators, functions, constants);
            return buildAst(tokens, operators, functions);
        },

        generateAST(expr) {
            const operators = new Set(Object.keys(OPERATORS));
            const functions = new Set(Object.keys(FUNCTIONS));
            const constants = new Set(Object.keys(CONSTANTS));

            const { expr: processedExpr, operators: sortedOperators } = preprocess(expr, operators, functions);
            const tokens = tokenize(processedExpr, sortedOperators, functions, constants);
            const ast = buildAst(tokens, operators, functions);
            
            const { nodes, edges } = generateMermaidAST(ast);
            return `graph TD;\n${nodes}${edges}`;
        }
    };
})();

export default Calculator;

// 使用示例
if (typeof window === 'undefined') {
    const expressions = [
        "max(-1,-1-1,1, min(10, -1-1, 2))",
        "-1",
        "1 + 2",
        "2 * (3 + 4)",
        "(1 + 2) * (-3 - 4)",
        "max(1, 2, 3)",
        "1# + 2",
        "1#",
        "150°",
        "deg(150)",
        "rad(PI/2)",
        "e#"
    ];

    for (const expr of expressions) {
        try {
            console.log(`\n处理表达式: ${expr}`);
            const result = Calculator.calculate(expr);
            console.log(`结果: ${result}`);
        } catch (e) {
            console.log(`错误: ${e.message}`);
        }
    }
} 

File: calculator.py
from graphviz import Digraph
import os
from operators import OPERATORS, FUNCTIONS

class Node:
    def __init__(self, value):
        self.value = value
        self.args = []

class Calculator:
    def __init__(self):
        self.operators = OPERATORS
        self.functions = set(FUNCTIONS.keys())
    
    def parse_expr(self, expr):
        tokens = self._tokenize(expr)
        return self._build_ast(tokens)
    
    def _tokenize(self, expr):
        """将表达式转换为标记列表"""
        tokens = []
        i = 0
        expr = expr.replace(' ', '')
        
        while i < len(expr):
            char = expr[i]
            
            # 处理数字
            if char.isdigit():
                num = ''
                while i < len(expr) and (expr[i].isdigit() or expr[i] == '.'):
                    num += expr[i]
                    i += 1
                tokens.append(('number', float(num)))
                continue
                
            # 处理函数名和括号
            if char.isalpha():
                func = ''
                while i < len(expr) and expr[i].isalpha():
                    func += expr[i]
                    i += 1
                if func in self.functions:
                    tokens.append(('function', func))
                continue
                
            # 处理运算符和括号
            if char in '+-*/#()':
                tokens.append(('operator', char))
            elif char == ',':
                tokens.append(('separator', char))
                
            i += 1
            
        return tokens
    
    def _build_ast(self, tokens):
        """使用调度场算法构建AST"""
        output = []  # 输出队列
        operators = []  # 运算符栈
        
        for token_type, token in tokens:
            if token_type == 'number':
                output.append(Node(token))
            
            elif token_type == 'function':
                operators.append(token)
            
            elif token_type == 'operator':
                if token == '(':
                    operators.append(token)
                elif token == ')':
                    while operators and operators[-1] != '(':
                        self._process_operator(operators, output)
                    if operators:  # 弹出左括号
                        operators.pop()
                        if operators and operators[-1] in self.functions:
                            self._process_operator(operators, output)
                else:  # 其他运算符
                    while (operators and operators[-1] != '(' and
                           operators[-1] in self.operators and
                           self.operators[operators[-1]]['precedence'] >= 
                           self.operators[token]['precedence']):
                        self._process_operator(operators, output)
                    operators.append(token)
        
        # 处理剩余的运算符
        while operators:
            self._process_operator(operators, output)
            
        return output[0] if output else None
    
    def _process_operator(self, operators, output):
        """处理运算符"""
        op = operators.pop()
        node = Node(op)
        
        if op in self.operators:
            args_count = self.operators[op]['args']
            node.args = [output.pop() for _ in range(args_count)]
            node.args.reverse()
        elif op in self.functions:
            # 收集函数参数直到遇到分隔符
            args = []
            while output and output[-1].value != ',':
                args.append(output.pop())
            node.args = list(reversed(args))
            
        output.append(node)

    def evaluate(self, node):
        if node is None:
            return 0
            
        # 如果是数字节点
        if isinstance(node.value, (int, float)):
            return node.value
            
        # 计算子节点
        args = [self.evaluate(arg) for arg in node.args]
        
        # 处理运算符
        if node.value in self.operators:
            return self.operators[node.value]['func'](*args)
        
        # 处理函数
        if node.value in self.functions:
            return FUNCTIONS[node.value]['func'](*args)
        
        raise ValueError(f"未知的操作符或函数: {node.value}")

    def visualize_ast(self, node, output_file='ast'):
        """
        将抽象语法树可视化为图形
        
        Args:
            node: AST的根节点
            output_file: 输出文件名（不需要扩展名）
        """
        dot = Digraph(comment='Abstract Syntax Tree')
        dot.attr(rankdir='TB')
        
        def add_nodes(node, parent_id=None, counter=[0]):
            if node is None:
                return
            
            # 为每个节点创建唯一ID
            node_id = str(counter[0])
            counter[0] += 1
            
            # 设置节点样式
            if isinstance(node.value, (int, float)):
                # 数值节点使用椭圆形
                dot.node(node_id, str(node.value), shape='ellipse')
            elif node.value in self.operators:
                # 运算符节点使用矩形
                dot.node(node_id, node.value, shape='box')
            elif node.value in self.functions:
                # 函数节点使用菱形
                dot.node(node_id, node.value, shape='diamond')
            else:
                # 其他节点使用默认形状
                dot.node(node_id, str(node.value))
            
            # 如果有父节点，添加边
            if parent_id is not None:
                dot.edge(parent_id, node_id)
            
            # 递归处理子节点
            for arg in node.args:
                add_nodes(arg, node_id)
        
        add_nodes(node)
        
        # 保存图形
        dot.render(output_file, view=True, format='png', cleanup=True)

def calculate(expr):
    calc = Calculator()
    tree = calc.parse_expr(expr)
    return calc.evaluate(tree)

# 测试代码
if __name__ == "__main__":
    expressions = [
        "1 + 2",
        "2 * (3 + 4)",
        "(1 + 2) * (3 + 4)",
        "max(1, 2, 3)",
        "1# + 2",
    ]
    
    calc = Calculator()
    for i, expr in enumerate(expressions):
        try:
            print(f"\n处理表达式: {expr}")
            tree = calc.parse_expr(expr)
            result = calc.evaluate(tree)
            print(f"结果: {result}")
            
            # 生成语法树可视化
            filename = f"ast_{i}"
            calc.visualize_ast(tree, filename)
            input("按任意键继续...")
             # 查看完后删除
            os.remove(f"{filename}.png")
            
        except Exception as e:
            print(f"错误: {str(e)}") 